sdk-core/src/main/java/care/data4life/sdk/fhir/Fhir4RecordClient.kt

```diff
@@ -75,4 +87,21 @@ internal class Fhir4RecordClient(
                 }
         return handler.executeSingle(operation, callback)
     }
+
+    override fun <T : Fhir4Resource> count(
```

Comment:
on iOS we have a usable version of count with the same signature that doesnt pass the resourceType (uses a default argument of Fhir4Resource). So its not nullable but its omittable either way.  is it possible to create the same here?

Reply:
Maybe lets split the concerns, and lets see iff I got them correctly:
1. is `resourceType` nullable? - no it is not (since in Kotlin you have to declare nullable values explicitly). Hence consumer has to delegate a certain value. 
2. make `resourceType` omittable. Yeah we could do that. But the question is - what is a reasonable default value? `DomainResource` maybe, but this would lead to a countAll behaviour, which we discussed in a other place.

Reply:
I mean, from API parity prospective, on the public interface of iOS we have 3 count methods (one for each case, fhir3/4 and appdata)  and the fhir ones look like:

```
public func countFhirStu3Records<R: FhirStu3Resource>(of type: R.Type = R.self,
                                                          annotations: [String] = [],
                                                          queue: DispatchQueue = responseQueue,
                                                          completion: @escaping ResultBlock<Int>) {
        fhirService
            .countFhirRecords(of: type, annotations: annotations)
            .complete(queue: queue, completion)
    }
```

so if someone wants to count all fhirStu3 records can just write:
`countFhirStu3Records { count in ... }`  and it automatically uses R.self as type (which is DomainResource, which recordService turns into a no-parameter, effectively doing a countAllFhir3Records)

I would like the android api to have the same structure if possible

Reply:
Ah...I think I got it new... `{...}` calls the completion, right?
Technically we can do it in the future, but for now we are bounded to the legacy client. I agree that our API should be aligned as possible crossplattform. However I do not see it in this particular task. Also the team did not decided how to handle the parallelisms in the future. For example to make something similar in Kotlin we need to change the SDKContract that the callback is a closure and not a `Callback`.

Reply:
@wmontwe maybe you have opinion as well? Its about implementing the count all behaviour via default parameter. What you like prefer more having a dedicated function `countAllSomething` or just overload `countSomething`.
The overload could or an alias `countAllSomething` could be added here.
I guess that has also some implication for a search all for `Something`.

* replace `Something` with a concrete FHIR type or with Data

Reply:
Hmm, I think self reference of a Generic type is not possible with Kotlin. Maybe I'm wrong. Also we need to ensure Java interop, which would require moving the parameter at last position to be able to create an overloaded function.

But I like both approaches:
* `countAll` is more intuitive, but broader API
* `count` with defaulting to DomainResource is slick, but maybe not as easy to grasp

`countAll` still leaves us with a client using `count` with `DomainResource`

Reply:
@ouafae31